File: src/Sentinel.Domain/Entities/LogEntry.cs
using Sentinel.Domain.Common;
using Sentinel.Domain.Enums;
using Sentinel.Domain.Events;

namespace Sentinel.Domain.Entities;

public class LogEntry : AggregateRoot
{
    public Guid AgentId { get; private set; }
    public string Source { get; private set; } = null!;
    public LogLevel Level { get; private set; }
    public string Message { get; private set; } = null!;
    public string? StackTrace { get; private set; }
    public Dictionary<string, object> Properties { get; private set; } = new();
    public DateTime Timestamp { get; private set; }

    public Agent Agent { get; private set; } = null!;

    private LogEntry() { }

    public static LogEntry Create(
        Guid agentId,
        string source,
        LogLevel level,
        string message,
        DateTime timestamp,
        string? stackTrace = null,
        Dictionary<string, object>? properties = null)
    {
        if (agentId == Guid.Empty)
            throw new ArgumentException("AgentId cannot be empty.", nameof(agentId));

        if (string.IsNullOrWhiteSpace(message))
            throw new ArgumentException("Message cannot be empty.", nameof(message));

        var logEntry = new LogEntry
        {
            AgentId = agentId,
            Source = source,
            Level = level,
            Message = message,
            Timestamp = timestamp,
            StackTrace = stackTrace,
            Properties = properties ?? new Dictionary<string, object>()
        };

        if (logEntry.IsCritical())
        {
            logEntry.AddDomainEvent(new CriticalLogReceivedEvent(
                logEntry.Id,
                logEntry.AgentId,
                logEntry.Timestamp));
        }

        return logEntry;
    }

    public bool IsCritical() => Level >= LogLevel.Error;
}

File: src/Sentinel.Mobile/lib/models/log_entry.dart
class LogEntry {
  final String id;
  final String agentId;
  final String agentName;
  final String source;
  final String level;
  final String message;
  final String? stackTrace;
  final Map<String, dynamic> properties;
  final String timestamp;

  LogEntry({
    required this.id,
    required this.agentId,
    required this.agentName,
    required this.source,
    required this.level,
    required this.message,
    this.stackTrace,
    required this.properties,
    required this.timestamp,
  });

  factory LogEntry.fromJson(Map<String, dynamic> json) {
    return LogEntry(
      id: json['id'] ?? '',
      agentId: json['agentId'] ?? '',
      agentName: json['agentName'] ?? '',
      source: json['source'] ?? '',
      level: json['level'] ?? 'Information',
      message: json['message'] ?? '',
      stackTrace: json['stackTrace'],
      properties: Map<String, dynamic>.from(json['properties'] ?? {}),
      timestamp: json['timestamp'] ?? '',
    );
  }
}
